### ISSUE TITLE
Feature request: Online backup (like Google Drive)

### ISSUE THREAD

**Issue by siddhpant** [mollyim/mollyim-android#224]
### Is there an existing request for this?

- [X] I have searched the existing issues

### Feature description

Currently, one of the major blocker for adoption of Signal is the database not being backed up online. If your phone is gone or reset, your messages are gone. This is not good for general usage.

Molly improves upon Signal client so that database can be passphrase protected at rest. Thus, it is possible to securely upload to online cloud storage without much hassle in restoring it later.

WhatsApp allows online backup to Google Drive. They have even introduced passphrase to encrypt your database, so only encrypted database is sent over the wire, and neither Google nor WhatsApp will have access to messages (as they say).

I request for the same feature in Molly.

---

Applications can upload their own files to Google Drive which cannot be accessed by anyone else (https://developers.google.com/drive/api/guides/appdata).

For FOSS version, maybe Nextcloud could be supported, or not support the functionality at all.

**Comment by siddhpant** [comment]
Well, I must say I'm not much into app dev, but I think upload to Google Drive is done using the free API whose link I posted above. The storage consumed is of the user account. And otherwise, maybe allowing to set the endpoint to upload would be a good option.

Yes, you didn't lose your chats, yet. Of course, you probably have good backup practices, but you cannot expect that from the average user.

**Comment by PsyNyde** [comment]
why not just use onedrive for syncing to cloud. Google drive also has some 3rd party software to do that also afaik.

**Comment by jcz1** [comment]
FYI: https://github.com/seedvault-app/seedvault/wiki/FAQ
(Molly seems to be backed-up by Seedvault)

**Comment by kasia125** [comment]
> Currently, one of the major blocker for adoption of Signal is the database not being backed up online. If your phone is lost or reset, your messages are gone. This is not good for general usage, especially considering "clear app data" is a generic troubleshooting step.
> 
> Molly improves upon Signal client so that database can be passphrase protected at rest. Thus, it is possible to securely upload to online cloud storage without much hassle in restoring it later.

This would require reducing security from the status quo implemented by Molly.

It doesn't matter whether it's on Google Drive or on your disk, but what would matter is how you decrypt it once you have a new phone after having lost your old one. See here for how Molly stores encryption secrets, and how decrpytion and encryption are handled: https://github.com/mollyim/mollyim-android/wiki/Data-Encryption-At-Rest
Currently, for some devices, Molly uses the Secure Element (SE) to store some cryptographic material, and, as written, "[this] key never leaves the chip." SEs and related hardware are (usually) designed to prevent or inhibit exporting cryptographic material from the hardware, but this comes with an advantage: if the material could be pulled from your device, then a well-resources adversary could extract it, then copy it to a more powerful computer. That might be a supercomputer, a rig of GPUs, or to specialized hardware, then they can decrypt your databases significantly faster than your SE could have.

Most importantly: your Molly database cannot be decrypted without access to your device. If Google obtains a copy of your database right now, they have no hope of decrypting your database, unless get physical access to your device, or if they run malware on your device that keeps trying to decrypt in in the background.

If online encrypted backups are to be useful, then this would have to be disabled. Having the option would encourage lowering security.

Maybe I misunderstood you? Do you just want to upload the files from the regular backup process built into Signal/Molly? Those backups aren't encrypted with a passphrase.

**Comment by siddhpant** [comment]
@kasia125 Yes, you have misunderstood everything.

You are just describing the regular storage of data which is encrypted at rest using the system you mentioned. That's not really an app specific feature, it's just how things are on Android.

The link you posted is telling you that on older Android versions where Keystore isn't used, the key is stored in plaintext alongside the database by the regular Signal app. So Molly encrypts that file with user passphrase, which is the feature in that context, so it doesn't matter whether Android saves the key in plaintext or other form.

I'm talking about the backup feature in `Settings -> Chat -> Backup`. Molly encrypts that file with the password you have set.

![Screenshot_20250216-181605_Molly.png](https://github.com/user-attachments/assets/273e04fe-8887-4f40-8636-c23224e284d0)

**Comment by siddhpant** [comment]
> why not just use onedrive for syncing to cloud. Google drive also has some 3rd party software to do that also afaik.

@psynyde Yeah I backup using a third party app as of now. But an inbuilt functionality would be great.

(Sorry for replying after a year 😅)

**Comment by kasia125** [comment]
> [@kasia125](https://github.com/kasia125) Yes, you have misunderstood everything.
> 
> You are just describing the regular storage of data which is encrypted at rest using the system you mentioned. That's not really an app specific feature, it's just how things are on Android.

No, it's both. Android does encrypt the disk with a key derived from your lock screen password, but Molly also separately uses the SE.

> I'm talking about the backup feature in `Settings -> Chat -> Backup`. Molly encrypts that file with the password you have set.

No these aren't encrypted with your password. Turn it off then turn it back on again; Molly will give you a key of multiple digits, I believe 15-25 digits.

**Comment by siddhpant** [comment]
> No, it's both. Android does encrypt the disk with a key derived from your lock screen password, but Molly also separately uses the SE.

Usage of SE doesn't increase any encryption. Its use, just like the encryption of shared preferences, is redundant and frankly shouldn't be there. Use of SE is only to protect the user's passphrase from brute force.

> No these aren't encrypted with your password. Turn it off then turn it back on again; Molly will give you a key of multiple digits, I believe 15-25 digits.
> 
> But yes, uploading the backups is fine. The backups are encrypted with a strong key / "passphrase."

Oh come on, stop being the grammar nazi here. You were clearly wrong, just own up to it.

**Comment by kasia125** [comment]
I genuinely did not know that Signal described this as encrypted with a passphrase

![Chat backups page in Molly](https://github.com/user-attachments/assets/71aef98a-7651-40ae-af3b-bfddf8c8ecb4)
![Chat backups page in Molly, with a prompt asking, "Enable local backups?" Along with "You must have this pasphrase in order to restore a backup."](https://github.com/user-attachments/assets/9a162cbc-c372-4bf7-8afc-2ba693ef14cf)

I just wouldn't have considered it a phrase, as it's numerical. I had suspected that it may have been the key, but it makes sense that it's not the raw key, because 30 decimal digits is only approximately 99 bits of entropy. log2(10^30) ~ 99.66

But I suppose I am being the grammar nazi here, I own up to that.

**Comment by kasia125** [comment]
Maybe the ability to automatically "share" the backup file would work?

It looks like Signal does not allow selecting a folder outside of the internal storage, otherwise you might be able to select Google Drive in the Files app if they exposed it there. This might be an Android limitation.

**Comment by HaleTom** [comment]
Note that Signal has cloud backups in testing.  It seems they're trying tom monetise it.

https://community.signalusers.org/t/public-signal-backups-testing/69984

It would be great if this project could use their open source code and allow people to use Google Drive storage instead.

**Comment by p1gp1g** [comment]
The new "cloud backup" (incremental backups) is finally available for on-device backups with v8.7.3-1. You can use a [SAF](https://developer.android.com/guide/topics/providers/document-provider) folder (eg. a WebDAV mount point using DAVx5 or another cloud application) for location to your backup, giving online backup capabilities.

Settings > Backups > On-device backups
